--- OC_User.php	2022-05-22 17:38:32.827000000 -0500
+++ OC_User.patched.php	2022-05-22 17:42:34.358000000 -0500
@@ -176,11 +176,15 @@
 					$message = \OC::$server->getL10N('lib')->t('User disabled');
 					throw new LoginException($message);
 				}
 				$userSession->setLoginName($uid);
 				$request = OC::$server->getRequest();
-				$userSession->createSessionToken($request, $uid, $uid);
+				$password = null;
+				if ($backend instanceof \OCP\Authentication\IProvideUserSecretBackend) {
+					$password = $backend->getCurrentUserSecret();
+				}
+				$userSession->createSessionToken($request, $uid, $uid, $password);
 				// setup the filesystem
 				OC_Util::setupFS($uid);
 				// first call the post_login hooks, the login-process needs to be
 				// completed before we can safely create the users folder.
 				// For example encryption needs to initialize the users keys first
@@ -188,11 +192,11 @@
 				OC_Hook::emit(
 					'OC_User',
 					'post_login',
 					[
 						'uid' => $uid,
-						'password' => null,
+						'password' => $password,
 						'isTokenLogin' => false,
 					]
 				);
 				/** @var IEventDispatcher $dispatcher */
 				$dispatcher = \OC::$server->get(IEventDispatcher::class);
