--- AppData.php.orig	2022-03-29 16:47:16.728000000 -0500
+++ AppData.php	2022-03-29 16:55:29.401000000 -0500
@@ -36,6 +36,7 @@
 use OCP\Files\NotFoundException;
 use OCP\Files\NotPermittedException;
 use OCP\Files\SimpleFS\ISimpleFolder;
+use OCP\Files\Storage\IStorageFactory;
 
 class AppData implements IAppData {
 
@@ -54,6 +55,9 @@
 	/** @var (ISimpleFolder|NotFoundException)[]|CappedMemoryCache */
 	private $folders;
 
+	/** @var string */
+	private $appdataRoot = "";
+
 	/**
 	 * AppData constructor.
 	 *
@@ -68,6 +72,19 @@
 		$this->config = $systemConfig;
 		$this->appId = $appId;
 		$this->folders = new CappedMemoryCache();
+
+		if (in_array('appdataroot',\OC::$server->getSystemConfig()->getKeys())) {				
+			$arguments = [
+				'datadir' => \OC::$server->getSystemConfig()->getValue('appdataroot'),
+			];
+			
+			$storage = new \OC\Files\Storage\LocalRootStorage($arguments);
+			$loader = \OC::$server->query(IStorageFactory::class);
+			$mount = new \OC\Files\Mount\MountPoint($storage, '/appdata/', $arguments, $loader);
+
+			\OC::$server->getMountManager()->addMount($mount);
+			$this->appdataRoot = 'appdata/';
+		}
 	}
 
 	private function getAppDataFolderName() {
@@ -80,11 +97,11 @@
 	}
 
 	protected function getAppDataRootFolder(): Folder {
-		$name = $this->getAppDataFolderName();
+		$name = $this->appdataRoot . $this->getAppDataFolderName();
 
 		try {
 			/** @var Folder $node */
-			$node = $this->rootFolder->get($name);
+			$node = $this->rootFolder->get($name);				
 			return $node;
 		} catch (NotFoundException $e) {
 			try {
@@ -101,7 +118,7 @@
 	 */
 	private function getAppDataFolder(): Folder {
 		if ($this->folder === null) {
-			$name = $this->getAppDataFolderName();
+			$name = $this->appdataRoot . $this->getAppDataFolderName();
 
 			try {
 				$this->folder = $this->rootFolder->get($name . '/' . $this->appId);
@@ -137,7 +154,7 @@
 			if ($name === '/') {
 				$node = $this->getAppDataFolder();
 			} else {
-				$path = $this->getAppDataFolderName() . '/' . $this->appId . '/' . $name;
+				$path = $this->appdataRoot . $this->getAppDataFolderName() . '/' . $this->appId . '/' . $name;
 				$node = $this->rootFolder->get($path);
 			}
 		} catch (NotFoundException $e) {
@@ -145,6 +162,14 @@
 			throw $e;
 		}
 
+		try {
+			$this->getAppDataRootFolder()->newFolder($key);
+			$this->getAppDataRootFolder()->delete($key);
+			throw new NotFoundException;
+		} catch (NotPermittedException $e) {
+			// Do nothing
+		}
+
 		/** @var Folder $node */
 		$folder = new SimpleFolder($node);
 		$this->folders->set($key, $folder);
@@ -178,4 +203,4 @@
 	public function getId(): int {
 		return $this->getAppDataFolder()->getId();
 	}
-}
\ No newline at end of file
+}
