--- AppData.php	2022-05-22 17:38:22.593000000 -0500
+++ AppData.patched.php	2022-05-22 17:41:50.504000000 -0500
@@ -34,10 +34,11 @@
 use OCP\Files\IRootFolder;
 use OCP\Files\Node;
 use OCP\Files\NotFoundException;
 use OCP\Files\NotPermittedException;
 use OCP\Files\SimpleFS\ISimpleFolder;
+use OCP\Files\Storage\IStorageFactory;
 
 class AppData implements IAppData {
 
 	/** @var IRootFolder */
 	private $rootFolder;
@@ -52,10 +53,13 @@
 	private $folder;
 
 	/** @var (ISimpleFolder|NotFoundException)[]|CappedMemoryCache */
 	private $folders;
 
+	/** @var string */
+	private $appdataRoot = "";
+
 	/**
 	 * AppData constructor.
 	 *
 	 * @param IRootFolder $rootFolder
 	 * @param SystemConfig $systemConfig
@@ -66,10 +70,23 @@
 								string $appId) {
 		$this->rootFolder = $rootFolder;
 		$this->config = $systemConfig;
 		$this->appId = $appId;
 		$this->folders = new CappedMemoryCache();
+
+		if (in_array('appdataroot',\OC::$server->getSystemConfig()->getKeys())) {		
+			$arguments = [
+				'datadir' => \OC::$server->getSystemConfig()->getValue('appdataroot'),
+			];
+			// Have to use `appdata_[instanceid]` as the root because of `OC\Encryption\Util->getUidAndFilename()`
+			$storage = new \OC\Files\Storage\LocalRootStorage($arguments);
+			$loader = \OC::$server->query(IStorageFactory::class);
+			$mount = new \OC\Files\Mount\MountPoint($storage, '/appdata_' . \OC::$server->getSystemConfig()->getValue('instanceid') . '/', $arguments, $loader);
+
+			\OC::$server->getMountManager()->addMount($mount);
+			$this->appdataRoot = 'appdata_' . \OC::$server->getSystemConfig()->getValue('instanceid') . '/';
+		}
 	}
 
 	private function getAppDataFolderName() {
 		$instanceId = $this->config->getValue('instanceid', null);
 		if ($instanceId === null) {
@@ -78,11 +95,11 @@
 
 		return 'appdata_' . $instanceId;
 	}
 
 	protected function getAppDataRootFolder(): Folder {
-		$name = $this->getAppDataFolderName();
+		$name = $this->appdataRoot . $this->getAppDataFolderName();
 
 		try {
 			/** @var Folder $node */
 			$node = $this->rootFolder->get($name);
 			return $node;
@@ -99,11 +116,11 @@
 	 * @return Folder
 	 * @throws \RuntimeException
 	 */
 	private function getAppDataFolder(): Folder {
 		if ($this->folder === null) {
-			$name = $this->getAppDataFolderName();
+			$name = $this->appdataRoot . $this->getAppDataFolderName();
 
 			try {
 				$this->folder = $this->rootFolder->get($name . '/' . $this->appId);
 			} catch (NotFoundException $e) {
 				$appDataRootFolder = $this->getAppDataRootFolder();
@@ -135,18 +152,31 @@
 		try {
 			// Hardening if somebody wants to retrieve '/'
 			if ($name === '/') {
 				$node = $this->getAppDataFolder();
 			} else {
-				$path = $this->getAppDataFolderName() . '/' . $this->appId . '/' . $name;
+				$path = $this->appdataRoot . $this->getAppDataFolderName() . '/' . $this->appId . '/' . $name;
 				$node = $this->rootFolder->get($path);
 			}
 		} catch (NotFoundException $e) {
 			$this->folders->set($key, $e);
 			throw $e;
 		}
 
+		$mountPoint = $this->getAppDataFolder()->getStorage()->getMountPoint();
+
+		if (! \OC::$server->getMountManager()->find($mountPoint)->getStorage()->file_exists($this->getAppDataFolderName() . '/' . $key)) {
+			$this->getAppDataRootFolder()->newFolder($key);
+		}
+//		try {
+//			$this->getAppDataRootFolder()->newFolder($key);
+//			$this->getAppDataRootFolder()->delete($key);
+//			throw new NotFoundException;
+//		} catch (NotPermittedException $e) {
+//			// Do nothing
+//		}
+
 		/** @var Folder $node */
 		$folder = new SimpleFolder($node);
 		$this->folders->set($key, $folder);
 		return $folder;
 	}
